### DISCUSSION TITLE
[RFC] New component Space.Compact

### DISCUSSION THREAD

**Discussion by foryuki** [ant-design/ant-design#37521] [RFCs - archive]
## Summary
新增 Space.Compact 组件，表示一个紧凑排列的组，支持将表单类组件（如 Input、Select、Button）进行成组包裹，并逐步废弃和替换原 Input.Group 和 Button.Group 的用法。

### 使用 Space 让 Button 成组
> 效果类似官网 Demo 《[废弃的 Block 组](http://ant-design.antgroup.com/components/button-cn/#components-button-demo-legacy-group)》

```jsx
<Space.Compact>
  <Button type="primary">Button 1</Button>
  <Button type="primary">Button 2</Button>
  <Tooltip title="Tooltip">
    <Button type="primary" icon={<DownloadOutlined />} disabled />
  </Tooltip>
  <Tooltip title="Tooltip">
    <Button type="primary" icon={<DownloadOutlined />} />
  </Tooltip>
</Space.Compact>

```

![image](https://user-images.githubusercontent.com/6828924/189676679-c1f0db33-07bb-4d55-bc5b-0f9180de6e7b.png)

### 使用 Space.Compact 让 Input 成组
> 效果类似官网 Demo 《[输入框组合](https://ant-design.antgroup.com/components/input-cn/#components-input-demo-group)》

```jsx
<Space.Compact>
  <Input style={{ width: '20%' }} defaultValue="0571" />
  <Input style={{ width: '30%' }} defaultValue="26888888" />
</Space.Compact>
```

![image](https://user-images.githubusercontent.com/6828924/189677796-74393551-f9f3-46b5-9c8e-8b3be8751b5a.png)

### 使用 Space.Compact 让表单类组件成组
> 效果类似官网 Demo 《[输入框组合](https://ant-design.antgroup.com/components/input-cn/#components-input-demo-group)》

```jsx
<>
  <Space.Compact>
    <Input style={{ width: 'calc(100% - 200px)' }} defaultValue="https://ant.design" />
    <Button type="primary">Submit</Button>
  </Space.Compact>
  <Space.Compact>
    <Input
      style={{ width: 'calc(100% - 200px)' }}
      defaultValue="git@github.com:ant-design/ant-design.git"
    />
    <Button icon={<CopyOutlined />} />
  </Space.Compact>
</>
```

![image](https://user-images.githubusercontent.com/6828924/189678672-a1773524-c4bc-4f3d-b555-bad98f2e3d0c.png)

## Motivation

* 目前 Input.Group 和 Button.Group 在一定程度上存在功能重复的情况，且 Input.Group 的非 compact 模式与 Space 组件展示和 API 上功能重叠
* 从 API 的一致性上来说，Input.Group 支持 compact 与否，而 Button.Group 默认为 compact 模式，应统一收敛
* Input.Group 可用于包括 Select 等其他表单组件，该 API namespace 挂在 Input 上表意不清。

通过 Space.Compact 组件，支持原 Input.Group 和 Button.Group 等表单类组件的成组效果，替换 Dropdown.Button 和 `<Input addon />` 等成组（并合并边框）实现，面向未来其他类似的基础的组件成组需求时，可统一收敛 API 到 Space.Compact 组件中。

## Detailed Design

新增 `SpaceCompactItemContext`

```ts
// components/config-provider/SizeContext.tsx
type SizeType = 'small' | 'middle' | 'large' | undefined;

interface SpaceCompactItemContextType {
  size?: SizeType;
  isFirstSpaceItem?: boolean;
  isLastSpaceItem?: boolean;
}

export const SpaceCompactItemContext = React.createContext<SpaceCompactItemContextType>({});
```

Input/Button 组件消费对应 Context 时，会产生以下行为表现：
* 读取 size context 以定制当前组件的尺寸大小，优先级高于当前组件自身的 size 属性。
* 读取并消费 isFirstSpaceItem、isLastSpaceItem context，进而产生一些样式适配。

Space.Compact 组件下嵌套组件的样式适配，应该都实现在对应组件内部，相当于 Space.Compact 提供了一个 Context，而对应的组件则响应对应的 Context 产生关联的样式适配。其消费方式为生成对应的 className 进而具备适配性样式。

### API

Space.Compact API 如下
| 参数 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| size | 被包裹子组件尺寸 | `middle` | `large` \| `middle` \| `small` | - | 
| direction | 间距方向 | `vertical` \| `horizontal` | `horizontal` | - |

### 当 Button/Input 响应 SpaceCompactItemContext 时，可能会产生很长的 classname
以 Button 组件为例，可能会产生 `ant-btn-compact-space-item` 及 `ant-btn-compact-space-first-item`、`ant-btn-compact-space-last-item` 这么长的样式类名，当然这些都属于内部实现，“影响美观” 但并不是很影响。

## Alternatives

通过添加类似全局的 css 类名嵌套约定，如新增全局约定的类名 `.ant-space-compact`，这样可以避免 context 传递和类名过长。

```css
.ant-space-compact {
  .ant-button:first-child {
  	// ...
  }
}
```

## Adoption strategy
* 在 v4 和 v5 中实现 Space.Compact 组件

## How we teach this
* 在 v4 中为 `Button.Group` 增加废弃声明，建议用户通过使用 `<Space.Compact />` 替代，并提示在 v5 中正式移除 `Button.Group` API 支持
* 在 v5 中为 `Input.Group` 增加废弃声明，建议用户通过使用 `Space` 或 `Space.Compact` 替代，并提示在 v6 中将正式移除 `Input.Group` API 支持
* 在 v4 -> v5 codemod 中增加对应的迁移脚本支持

### 当通过 Context 传递 SpaceCompactItemContext 时，遇到嵌套很深的情况下，对应的组件是否应响应产生样式变更?

暂时不处理，因为 Button 嵌套 Button 的场景本身很奇怪。若之后需要处理，则考虑通过 Context 阻断的方式解决（Context.Provider value={initialValue} />，会有一些性能牺牲)。

```jsx
<Space.Compact>
  <Button>
    Button
    <Button>Button child</Button>
  </Button>
</Space.Compact>
```

### Space 套 Space 如何处理

待更新

### 对于现存的 Button.Group/Input.Group 的部分样式是否可以改进得更好？

* Button.Group 内 primary 边界亮线如何实现
![image](https://user-images.githubusercontent.com/6828924/189676679-c1f0db33-07bb-4d55-bc5b-0f9180de6e7b.png)

* Input.Group 第一个 input focus 时右侧边框不高亮
![image](https://user-images.githubusercontent.com/6828924/189677796-74393551-f9f3-46b5-9c8e-8b3be8751b5a.png)

**Comment by MadCcc** [comment]
感觉和 compact 模式语义上有点相似，但是实际上是 zero-gap 🤔

**Reply by zombieJ** [reply]
那 collapsed？

**Reply by vagusX** [reply]
现在 size=[0, 0] 是没有边框合并的，不如直接叫做 `border-collapse`

**Reply by zombieJ** [reply]
另外还有 `size` 现在是间距，所以这个属性最好可以一同用于配置内部元素大小

**Comment by afc163** [comment]
不如分离开，<Space.Compact>

- Space size 和 Compact size 语义不一致，`mode="compact"` 时的 size 会很奇怪。

**Comment by zombieJ** [comment]
`cloneElement` 不好，这样 HOC 组件容易失效。建议还是 context 告知（如果要做的严谨一些可以组件自身做一个截断）。

**Reply by MadCcc** [reply]
考虑一下 Space 套 Space 的情况？其他情况下非容器组件感觉可以不用截断

**Comment by afc163** [comment]
v5 就不要移除了，v5 的 break change 在做 v5 RoadMap 时应该已经确定了，不要加戏。建议 v6 移除。

**Reply by zombieJ** [reply]
同 `visible` 文档移除，v6 再下。

**Comment by afc163** [comment]
1. 垂直合并也能做 https://mui.com/zh/material-ui/react-button-group/#vertical-group
2. Dropdown.Button 这个可以用 Compact 包 Button 来代替了，本质上就是 Split Button https://mui.com/zh/material-ui/react-button-group/#split-button

**Comment by vagusX** [comment]
input addon 移除，换成新实现，addon 换上之后的样式（现在 addon 有灰色底的问题）
使用 typography.text 来实现
支持 direction 方向
Space 套 Space 怎么展示
`<Provider value={initialValue} />` 后期考虑
Button primary 边界亮线如何实现
input focus 时右侧边框不高亮

**Comment by dancerphil** [comment]
这个 RFC 和早期的一个设计模式有冲突，可以把设计模式的说明去除：https://ant.design/docs/spec/buttons-cn#%E6%8C%89%E9%92%AE%E7%BB%84

**Comment by foryuki** [comment]
## PR 见：https://github.com/ant-design/ant-design/pull/37652

## 实现了

### 横向紧凑布局边框折叠

目前支持的组件有：

- Button
- AutoComplete
- Cascader
- DatePicker
- Input/Input.Search
- Select
- TimePicker
- TreeSelect

### 纵向紧凑布局

纵向紧凑布局暂时只支持 Button 组件。

## 下一步计划

### Dropdown.Button 换成 Space.Compact 实现
PR 见： https://github.com/ant-design/ant-design/pull/38090


![image](https://user-images.githubusercontent.com/8649233/194505275-a59ceac9-a5af-44be-a218-83b4048982be.png)


### Input addon 换成 Space.Compact 实现
input 的 addon 需要在 Space.Compact 下做边框合并（等到 addon 换为 Space.Compact 再实现）。

![image](https://user-images.githubusercontent.com/8649233/194505320-5bb86bc5-0423-41a8-8ac7-f35becf95581.png)

**Comment by foryuki** [comment]
从这个 issue https://github.com/ant-design/ant-design/issues/38341 引申两个问题：

1. 当 item 高度不一致时，是否需要将高度拉伸到一致高度

当 item 高度跟标准 size 不一致时，如下图 input 被 addon 中 button 撑高了，所以导致右侧的 button 高度显得不一致，这时是否需要拉高到一致

<img width="459" alt="image" src="https://user-images.githubusercontent.com/8649233/200343609-bd5fb950-220a-42a2-afeb-6d76bb8dbcc5.png">

可以通过设置 `height: auto` 解决：https://github.com/ant-design/ant-design/pull/38376


2. 当不设置 Space.Comact size 时，是否需要支持单独设置每个 item 的 size？（目前 v4 上 Input.Group 是支持的。）

如果支持了就会有如下场景，但是以下场景又跟上面的 高度拉伸冲突，也就是 size 为 sm 的 input 会跟 size 为 large 的 input 高度一致，从而导致这里的 单独设置 item size 是一个伪命题
<img width="951" alt="image" src="https://user-images.githubusercontent.com/8649233/200340813-c4a8fb19-87f1-492e-8017-f36ca3f0f31e.png">

**Comment by ByYogi** [comment]
万恶之源是吧🧐
## 用Space.Compact遇到很多坑
* UI遇到了问题：https://github.com/ant-design/ant-design/issues/41876
* 在与Form.Item关系，执行数据操作遇到了问题：https://github.com/ant-design/ant-design/issues/42001
